### PR TITLE
Add TransactionEnable in Namespace

### PR DESCRIPTION
**Pull request by liangyepianzhou**
Add TransactionEnable in Namespace. There are some problems in namespaceTest and it is being debugged. Submit after debugging.

### CODE REVIEW

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 1]
```diff
@@ -482,9 +486,18 @@ private void checkCanRemovePendingAcksAndHandle(PositionImpl position, MessageId
     }
 
     private boolean isTransactionEnabled() {
+        TopicName topicName = TopicName.get(subscription.getTopicName());
+        Optional<Policies> policies = null;
+        try {
+            policies = cnx.getBrokerService().pulsar().getConfigurationCache().policiesCache()
+                    .get(AdminResource.path(POLICIES, topicName.getNamespaceObject().toString()));
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:495]: Please log error

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java [hunk 2]
```diff
@@ -482,9 +486,18 @@ private void checkCanRemovePendingAcksAndHandle(PositionImpl position, MessageId
...
+        }
         return subscription instanceof PersistentSubscription
                 && ((PersistentTopic) subscription.getTopic())
-                .getBrokerService().getPulsar().getConfig().isTransactionCoordinatorEnabled();
+                .getBrokerService().getPulsar().getConfig().isTransactionCoordinatorEnabled()
+                && policies.get().transaction_enable;
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/Consumer.java:500]: There may be NPE, because policies can be null

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java [hunk 3]
```diff
@@ -140,11 +143,21 @@ public PersistentSubscription(PersistentTopic topic, String subscriptionName, Ma
         this.fullName = MoreObjects.toStringHelper(this).add("topic", topicName).add("name", subName).toString();
         this.expiryMonitor = new PersistentMessageExpiryMonitor(topicName, subscriptionName, cursor, this);
         this.setReplicated(replicated);
+        TopicName topicName1 = TopicName.get(getTopicName());
+        Optional<Policies> policies = null;
+        try {
+            policies = topic.getBrokerService().pulsar().getConfigurationCache().policiesCache()
+                     .get(AdminResource.path(POLICIES, topicName1.getNamespaceObject().toString()));
+        } catch (Exception e) {
+            e.printStackTrace();
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:152]: Same as above

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java [hunk 4]
```diff
@@ -394,8 +416,16 @@ public void acknowledgeMessage(List<Position> positions, AckType ackType, Map<St
...
         }
-
+        TopicName topicName1 = TopicName.get(getTopicName());
+        Optional<Policies> policies = null;
+        try {
+            policies = topic.getBrokerService().pulsar().getConfigurationCache().policiesCache()
+                    .get(AdminResource.path(POLICIES, topicName1.getNamespaceObject().toString()));
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:426]: Should we abstract it as a method?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java [hunk 5]
```diff
@@ -343,7 +351,16 @@ public PersistentTopic(String topic, ManagedLedger ledger, BrokerService brokerS
...
+        Optional<Policies> policies = null;
+        try {
+            policies = brokerService.pulsar().getConfigurationCache().policiesCache()
+                    .get(AdminResource.path(POLICIES, topicName.getNamespaceObject().toString()));
+        } catch (Exception e) {
+            e.printStackTrace();
+        }
+        if (brokerService.pulsar().getConfiguration().isTransactionCoordinatorEnabled()
+                && policies.get().transaction_enable) {
             this.transactionBuffer = brokerService.getPulsar()
```
Comment (315157973) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:364]: If we do not create a transactionBuffer, will the client still send data to the TC?

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/NamespacesImpl.java [hunk 6]
```diff
@@ -3616,6 +3616,66 @@ public void clearProperties(String namespace) throws PulsarAdminException {
...
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new PulsarAdminException(e);
+        } catch (TimeoutException e) {
+            throw new PulsarAdminException.TimeoutException(e);
+        }
+    }
+
+    @Override
+    public CompletableFuture<Boolean> getTransactionEnabledAsync(String namespace) {
```
Comment (315157973) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/NamespacesImpl.java:3656]: We also need to add CMD

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java (hunk 4 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:426]: There are one or two places used in a class, and there is no meeting point in itself.Is it necessary to abstract it into a method？

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java (hunk 5 above)
Comment (liangyepianzhou) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentTopic.java:364]: In my humble opinion, transactionBuffer should look like topic level.

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/NamespacesImpl.java (hunk 6 above)
Comment (liangyepianzhou) [reply pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/NamespacesImpl.java:3656]: Excuse me， I didn't understand what you mean

File: conf/standalone.conf [hunk 7]
```diff
@@ -590,7 +590,7 @@ managedLedgerDefaultAckQuorum=1
 
 # How frequently to flush the cursor positions that were accumulated due to rate limiting. (seconds).
 # Default is 60 seconds
-managedLedgerCursorPositionFlushSeconds = 60
+managedLedgerCursorPositionFlushSeconds=60
```
Comment (eolivelli) [inline conf/standalone.conf:593]: this change looks unrelated, can you please revert?

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java [hunk 8]
```diff
@@ -1853,6 +1853,32 @@ public void removeNamespaceResourceGroup(@PathParam("tenant") String tenant,
         validateNamespaceName(tenant, namespace);
         internalSetNamespaceResourceGroup(null);
     }
+    @POST
+    @Path("/{tenant}/{namespace}/transactionEnabled")
+    @ApiOperation(value = "Update boolean of whether allow transaction of  namespace")
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java:1858]: "Enable or disable transactions on the given namespace"

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java [hunk 9]
```diff
@@ -1853,6 +1853,32 @@ public void removeNamespaceResourceGroup(@PathParam("tenant") String tenant,
...
+            @PathParam("tenant") String tenant,
+            @PathParam("namespace") String namespace,
+            @ApiParam(value = "boolean of whether allow transaction of  namespace", required = true)
+                    boolean transactionEnable) {
+        validateNamespaceName(tenant, namespace);
+        internalSetTransactionEnabled(transactionEnable);
+    }
+    @GET
+    @Path("/{tenant}/{namespace}/transactionEnabled")
+    @ApiOperation(value = "The boolean of whether allow transaction of  namespace")
```
Comment (eolivelli) [inline pulsar-broker/src/main/java/org/apache/pulsar/broker/admin/v2/Namespaces.java:1872]: "Return information about the activation of transactions on the given namespace"

File: pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java (hunk 3 above)
Comment (eolivelli) [reply pulsar-broker/src/main/java/org/apache/pulsar/broker/service/persistent/PersistentSubscription.java:152]: we must throw an error here, because we cannot proceed if we do not have this piece of information

File: pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/NamespacesImpl.java [hunk 10]
```diff
@@ -3616,6 +3616,66 @@ public void clearProperties(String namespace) throws PulsarAdminException {
...
+        NamespaceName ns = NamespaceName.get(namespace);
+        WebTarget path = namespacePath(ns, "transactionEnabled");
+        final CompletableFuture<Boolean> future = new CompletableFuture<>();
+        asyncGetRequest(path,
```
Comment (eolivelli) [inline pulsar-client-admin/src/main/java/org/apache/pulsar/client/admin/internal/NamespacesImpl.java:3660]: who does this work during an upgrade?
if I make the request to a broker that is still at 2.8.0 version ?

File: pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/Policies.java [hunk 11]
```diff
@@ -125,6 +125,8 @@
     @SuppressWarnings("checkstyle:MemberName")
     public String resource_group_name = null;
 
+    @SuppressWarnings("checkstyle:MemberName")
+    public boolean transaction_enable = false;
```
Comment (eolivelli) [inline pulsar-client-admin-api/src/main/java/org/apache/pulsar/common/policies/data/Policies.java:129]: This should be "true" by default, otherwise when you upgrade from Pulsar 2.8.0 to Pulsar 2.9.0 you will see Transactions disabled for every namespace.

**Review by eolivelli — CHANGES_REQUESTED** [review]
This is a very interesting work.

I have a concern about the upgrade process.

In my opinion when you update from Pulsar 2.8 to Pulsar 2.9 and you are using Transactions you must be able to continue to use them without a manual reconfiguration, otherwise you will break the system.

I have a proposal:
-let Policies.transaction_enable be "true" be default
- introduce a new broker level configuration "`namespaceEnableTransactionDefault`=true"
- when you create a new namespace you set explicitly  transation_enable = `namespaceEnableTransactionDefault`

if you want transactions disabled by default you set namespaceEnableTransactionDefault=false in your broker.conf
otherwise the behaviour is the same as in Pulsar 2.8.0

In Pulsar 2.8.0 we blessed Transactions to be "production ready" and people started to use them, we must support the upgrade (and the downgrade) path

**Review by sijie — CHANGES_REQUESTED** [review]
I don't think we should support enabling or disabling transactions on a per-namespace basis. This will screw up the transaction semantic.

**Comment by codelipenghui** [comment]
@sijie Oh, good point. One transaction might acrosses multiple namespaces. If parts of the namespaces are disabled transactions, this will screw up the transaction semantic.

**Comment by liangyepianzhou** [comment]
Because it will affect cross-namespace transactions, namespace-level transactions will not be added temporarily
